### ISSUE TITLE
undo file file name too long on linux

### ISSUE THREAD

**Issue by GoogleCodeExporter** [vim/vim#346]
```
What steps will reproduce the problem?
1. open file of which the full path+filename is longer than the maximum 
filename-length allowed by linux; in my case this is I think >134, because of 
cryptfs module. Otherwise >255
2. alter something, hit save

What is the expected output? What do you see instead?
Expected, normal write, no problem
In stead:
E828: Cannot open undo file for writing: 
/home/john/.vim/undodir/%home%john%projects%some_project%some_joomla_directory%c
omponents%com_jreviews%jreviews%cms_compat%joomla%models%everywhere%everywhere_c
om_content.php

What version of the product are you using? On what operating system?
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Jan  2 2014 19:39:59)
Ubunbu linux with cryptfs

Please provide any additional information below.
in .vimrc:
set undofile
set undodir=/home/john/.vim/undodir/

I found out it is the filename length because I tried creating the file 
manually, investigating what went wrong, and linux said the name was too long. 
On searching documentation and forums I have come up with zero information 
about this behaviour.

My suggestion for a clean fix for this is using a hash of path in stead of full 
path. 

```

Original issue reported on code.google.com by `verders...@gmail.com` on 22 Mar 2015 at 10:24

**Comment by diversemix** [comment]
I'd certainly be interested in this -- seems in issue on my `ext4` partition

**Comment by ApolloTang** [comment]
For a get-around temporary solution, I found this plugin:

https://github.com/pixelastic/vim-undodir-tree

**Comment by ApolloTang** [comment]
As suggested by [@ddickstein](https://github.com/ddickstein) in [https://github.com/neovim/neovim/issues/8137#issuecomment-830668534](https://github.com/neovim/neovim/issues/8137#issuecomment-830668534) this could possibly be solved by hashing the paths?

**Comment by chrisbra** [comment]
It's part of the [todo.txt](https://github.com/vim/vim/blob/05a627c3d4e42a18f76c14828d68ee4747118211/runtime/doc/todo.txt#L5409-L5410)

**Comment by brammool** [comment]
> As suggested by ***@***.***(https://github.com/ddickstein) in
> [https://github.com/neovim/neovim/issues/8137#issuecomment-830668534](https://github.com/neovim/neovim/issues/8137#issuecomment-830668534)
> this could possibly be solved by hashing the paths?

Yes, that is also what is mentioned in the todo list.  Although this
solves the length problem, it makes it impossible to tell what original
file an undo file is for.  Another suggested solution is to recreate the
directory structure below 'undodir'.  The disadvantage of that is
performance.  Both creating the directory structure and searching
through it add quite a bit of overhead (file systems tend to cache the
contents of a directory, not a directory tree).

-- 
Violators can be fined, arrested or jailed for making ugly faces at a dog.
		[real standing law in Oklahoma, United States of America]

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
